File: src/API/ontologenius/clientsIndex/ontologyClients/IndividualIndexClient.cpp
#include "ontologenius/API/ontologenius/clientsIndex/ontologyClients/IndividualIndexClient.h"

#include <cstdint>
#include <string>
#include <vector>

namespace onto {

  std::vector<int64_t> IndividualIndexClient::getOn(int64_t index, int64_t property, int64_t selector)
  {
    std::string param = std::to_string(index) + ":" + std::to_string(property);
    if(selector != 0)
      param += " -s " + std::to_string(selector);

    return callIndexes("getOn", param);
  }

  std::vector<int64_t> IndividualIndexClient::getFrom(int64_t property, int64_t index, int64_t selector)
  {
    std::string param = std::to_string(index) + ":" + std::to_string(property);
    if(selector != 0)
      param += " -s " + std::to_string(selector);

    return callIndexes("getFrom", param);
  }

  std::vector<int64_t> IndividualIndexClient::getWith(int64_t indiv_from, int64_t indiv_to, int64_t selector, int depth)
  {
    std::string param = std::to_string(indiv_from) + ":" + std::to_string(indiv_to);
    if(selector != 0)
      param += " -s " + std::to_string(selector);

    if(depth >= 0)
      param += " -d " + std::to_string(depth);

    return callIndexes("getWith", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelatedFrom(int64_t property)
  {
    const std::string param = std::to_string(property);
    return callIndexes("getRelatedFrom", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelatedOn(int64_t property)
  {
    const std::string param = std::to_string(property);
    return callIndexes("getRelatedOn", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelatedWith(int64_t index)
  {
    const std::string param = std::to_string(index);
    return callIndexes("getRelatedWith", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelationFrom(int64_t index, int depth)
  {
    std::string param = std::to_string(index);
    if(depth >= 0)
      param += " -d " + std::to_string(depth);

    return callIndexes("getRelationFrom", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelationOn(int64_t index, int depth)
  {
    std::string param = std::to_string(index);
    if(depth >= 0)
      param += " -d " + std::to_string(depth);

    return callIndexes("getRelationOn", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRelationWith(int64_t index)
  {
    const std::string param = std::to_string(index);
    return callIndexes("getRelationWith", param);
  }

  std::vector<int64_t> IndividualIndexClient::getDomainOf(int64_t index, int64_t selector, int depth)
  {
    std::string param = std::to_string(index);
    if(selector != 0)
      param += " -s " + std::to_string(selector);

    if(depth >= 0)
      param += " -d " + std::to_string(depth);

    return callIndexes("getDomainOf", param);
  }

  std::vector<int64_t> IndividualIndexClient::getRangeOf(int64_t index, int64_t selector, int depth)
  {
    std::string param = std::to_string(index);
    if(selector != 0)
      param += " -s " + std::to_string(selector);

    if(depth >= 0)
      param += " -d " + std::to_string(depth);

    return callIndexes("getRangeOf", param);
  }

  std::vector<int64_t> IndividualIndexClient::getType(int64_t index)
  {
    const std::string param = std::to_string(index);
    return callIndexes("getType", param);
  }

  std::vector<int64_t> IndividualIndexClient::getSame(int64_t index)
  {
    const std::string param = std::to_string(index);
    return callIndexes("getSame", param);
  }

  std::vector<int64_t> IndividualIndexClient::getDistincts(int64_t index)
  {
    const std::string param = std::to_string(index);
    return callIndexes("getDistincts", param);
  }

  bool IndividualIndexClient::isInferred(int64_t subject, int64_t property, int64_t object)
  {
    std::string param = std::to_string(subject) + "|" + std::to_string(property) + "|" + std::to_string(object);
    return (callStr("isInferred", param).empty() == false);
  }

  bool IndividualIndexClient::isInferred(int64_t subject, int64_t class_index)
  {
    std::string param = std::to_string(subject) + "|" + std::to_string(class_index);
    return (callStr("isInferred", param).empty() == false);
  }

  std::vector<std::string> IndividualIndexClient::getInferenceExplanation(int64_t subject, int64_t property, int64_t object)
  {
    std::string param = std::to_string(subject) + "|" + std::to_string(property) + "|" + std::to_string(object);
    return call("getInferenceExplanation", param);
  }

  std::vector<std::string> IndividualIndexClient::getInferenceExplanation(int64_t subject, int64_t class_index)
  {
    std::string param = std::to_string(subject) + "|" + std::to_string(class_index);
    return call("getInferenceExplanation", param);
  }

} // namespace onto

File: ontopy/ontologenius/clientsIndex/IndividualIndexClient.py
from .OntologyIndexClient import OntologyIndexClient

class IndividualIndexClient(OntologyIndexClient):
    """The IndividualIndexClient class provides an abstraction of ontologenius individuals ROS service.
       The ontologenius individuals' service allows the exploration of individuals contained by ontologenius core based on indexes.
       This class is based on ClientBaseIndex and so ensure a persistent connection with ontologenius/individual_index service.
       The persistent connection ensures a minimal response time.
       A reconnection logic is implemented in the event that the persistent connection fails. 
    """

    def __init__(self, name):
        """Constructs an class client.
           Can be used in a multi-ontology mode by specifying the name of the ontology name(str).
           For classic use, name should be defined as "".
        """
        if name == '':
            OntologyIndexClient.__init__(self, "individual_index")
        else:
            OntologyIndexClient.__init__(self, "individual_index/" + name)

    def getOn(self, index, property, selector = 0):
        """Gives all the individuals (integer[]) pointed by the property(integer) and applied to the individual index(integer).
           The optional selector(integer) parameter can be set to only get results inheriting from the selector class.
           The default value 0 represents no restriction on the result.
        """
        param = str(index) + ":" + str(property)
        if selector != '':
            param += " -s " + str(selector)
        return self.callIndexes("getOn", param)

    def getFrom(self, property, index, selector = 0):
        """Gives all the individuals (integer[]) having the given property(integer) and pointing to the individual index(integer).
           The optional selector(integer) parameter can be set to only get results inheriting from the selector class.
           The default value 0 represents no restriction on the result.
        """
        param = str(index) + ":" + str(property)
        if selector != '':
            param += " -s " + str(selector)
        return self.callIndexes("getFrom", param)

    def getWith(self, indiv_from, indiv_to, selector = 0, depth = -1):
        """Gives all the properties (integer[]) linking the individual indiv_from(integer) to the individual indiv_to(integer).
           The optional selector(integer) parameter can be set to only get results inheriting from the selector property.
           The default value 0 represents no restriction on the result.
           The optional depth(int) parameter can be set to limit tree propagation to a specific value.
           The default value -1 represents no propagation limitation.
        """
        param = str(indiv_from) + ":" + str(indiv_to)
        if selector != '':
            param += " -s " + str(selector)
        if depth >= 0:
            param += " -d " + str(depth)
        return self.callIndexes("getWith", param)

    def getRelatedFrom(self, property):
        """Gives all the individuals (integer[]) possessing the property(integer)."""
        return self.callIndexes("getRelatedFrom", str(property))

    def getRelatedOn(self, property):
        """Gives all the individuals (integer[]) pointed to by the property(integer)."""
        return self.callIndexes("getRelatedOn", str(property))

    def getRelatedWith(self, index):
        """Gives all the individuals (integer[]) having a property pointing to the individual index(integer)."""
        return self.callIndexes("getRelatedWith", str(index))

    def getRelationFrom(self, index, depth = -1):
        """Gives all the properties (integer[]) applied to the individual index(integer).
           The optional depth(int) parameter can be set to limit tree propagation to a specific value.
           The default value -1 represents no propagation limitation.
        """
        param = str(index)
        if depth >= 0:
            param += " -d " + str(depth)
        return self.callIndexes("getRelationFrom", param)

    def getRelationOn(self, index, depth = -1):
        """Gives all the properties (integer[]) going to the individual index(integer).
           The optional depth(int) parameter can be set to limit tree propagation to a specific value.
           The default value -1 represents no propagation limitation.
        """
        param = str(index)
        if depth >= 0:
            param += " -d " + str(depth)
        return self.callIndexes("getRelationOn", param)

    def getRelationWith(self, index):
        """Gives all the individuals (integer[]) pointed by a property applied to the individual index(integer)."""
        return self.callIndexes("getRelationWith", str(index))

    def getDomainOf(self, index, selector = 0, depth = -1):
        """Gives all the properties (integer[]) for which the individual index(integer) is part of the domain.
           The optional selector(integer) parameter can be set to only get results inheriting from the selector property.
           The default value 0 represents no restriction on the result.
           The optional depth(int) parameter can be set to limit tree propagation of the individual to a specific value.
           The default value -1 represents no propagation limitation.
        """
        param = str(index)
        if selector != '':
            param += " -s " + str(selector)
        if depth >= 0:
            param += " -d " + str(depth)
        return self.callIndexes("getDomainOf", param)

    def getRangeOf(self, index, selector = 0, depth = -1):
        """Gives all the properties (integer[]) for which the individual index(integer) is part of the range.
           The optional selector(integer) parameter can be set to only get results inheriting from the selector property.
           The default value 0 represents no restriction on the result.
           The optional depth(int) parameter can be set to limit tree propagation of the individual to a specific value.
           The default value -1 represents no propagation limitation.
        """
        param = str(index)
        if selector != '':
            param += " -s " + str(selector)
        if depth >= 0:
            param += " -d " + str(depth)
        return self.callIndexes("getRangeOf", param)

    def getType(self, index):
        """Gives all the individuals (integer[]) of the type of the given class index(integer)."""
        return self.callIndexes("getType", str(index))

    def getSame(self, index):
        """Gives all the individuals (integer[]) that are defined as being identical to the individual index(integer)."""
        return self.callIndexes("getSame", str(index))

    def getDistincts(self, index):
        """Gives all the defined individuals (integer[]) as being distinct from the individual index(integer)."""
        return self.callIndexes("getDistincts", str(index))
